### DISCUSSION TITLE
Various NuGet repository issues (V3, NuGet.config, registryUrls)

### DISCUSSION THREAD

**Discussion by dannyyy** [renovatebot/renovate#7912] [Request Help] [answered]
**What Renovate type, platform and version are you using?**

Self-hosted Docker (Tag: 23)

**There are various (possible) bugs**
1. If a Git repository have a `NuGet.config` then all the configurations in the `nuget` section inside the `renovate.js` are ignored. The expected behavior is not clear according to the doc. Is there a possibility to merge the repo URLs or to ignore the `Nuget.config` as it can be done for `npm` with `ignoreNpmrcFile: true`
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
    <add key="DG" value="https://<organization>.pkgs.visualstudio.com/_packaging/<project>/nuget/v3/index.json" />
  </packageSources>
</configuration>
```  
```
    nuget: {
        registryUrls: [
            'https://api.nuget.org/v3/index.json',
            'https://api.nuget-dummy.org/v3/index.json',
        ]
    }
```
2. Having only a `NuGet.config` and no `nuget` section inside the `renovate.js`, renovate is ignoring the public NuGet repo `https://api.nuget.org/v3/index.json` and always try to query the private one for everything. The log does not mention any try to the public.

3. Having only a `nuget` section inside the `renovate.js`, the public repo is still ignored. Same as mentioned under `2.`. It doesn't matter whether the repo is specified with `https://api.nuget.org/v3/index.json` or `https://api.nuget.org/v3/index.json#protocolVersion=3`  
Only defining an old public repository `https://www.nuget.org/api/v2/` seems to work next to the private one. But usually an existing `Nuget.config` will never ever use this old interface.

4. Setting `enabled: false` to a `hostRule` of `hostType: 'nuget'` has no effect

Full `renovate.js` config
```
module.exports = {
    endpoint: 'https://organization@dev.azure.com/organization/',
    token: '...',
    gitAuthor: 'Renovate <renovate@organization.com>',
    onboarding: false,
    requireConfig: true,
    extends: [
        'config:base',
        ':prHourlyLimitNone'
    ],
    platform: 'azure',
    enabledManagers: ["npm", "nuget"],
    ignoreNpmrcFile: true,
    npmrc: '...',
    nuget: {
        registryUrls: [
            'https://api.nuget.org/v3/index.json',
            'https://organization.pkgs.visualstudio.com/_packaging/MyRepo/nuget/v3/index.json',
        ]
    },
    hostRules: [
        {
            hostType: 'npm',
            domainName: 'pkgs.dev.azure.com',
            username: 'organization',
            password: '...'
        },
        {
            hostType: 'npm',
            domainName: 'organization.pkgs.visualstudio.com',
            username: 'organization',
            password: '...'
        },
        {
            hostType: 'nuget',
            endpoint: 'https://organization.pkgs.visualstudio.com',
            username: '',
            password: '...',
        },
        {
            domain: 'github.com',
            token: '...'
        }
    ],
    repositories: ['Playground/MyGitRepo']
};
```

Full `Nuget.config` config
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
    <add key="DG" value="https://organization.pkgs.visualstudio.com/_packaging/MyRepo/nuget/v3/index.json" />
  </packageSources>
</configuration>
```

**Relevant debug logs**
Attached as [renovate.log](https://github.com/renovatebot/renovate/files/5660770/renovate.log). To follow up the behavior full log must be attached. The relevant lines including all the duplications (don't know why this is necessary) show the reported behavior. But unfortunately, not the reason why. 

**To Reproduce**

Run it on any Azure DevOps environment with the configs given. It's 100% reproduceable. 

**Additional context**

I guess, either the must be a configuration error, or the business logic behind is buggy. Funnily, the trace-log is much less verbose and helpful than debug. The trace output only duplicates nonsense-logging dozens of time.

**Comment by viceice** [comment, marked as answer]
@dannyyy Please try the following package-rule instead of  the `nuget` object.

```js
  packageRules: [
    {
      datasources: [ 'nuget' ],
      registryUrls: [
            'https://api.nuget.org/v3/index.json',
            'https://api.nuget-dummy.org/v3/index.json',
      ]
    }
  ]
```

this should hopefully override the extracted registry urls

**Reply by fdw** [reply]
Is this still valid? It doesn't seem to work for me (at least not reliably) and according to https://docs.renovatebot.com/renovate-schema.json , there should be no `registryUrls`.

**Comment by viceice** [comment]
also please checkout [hostRules](https://docs.renovatebot.com/configuration-options/#hostrules) docs. `endpoint` is wrong, maybe you mean `baseUrl` ? `domain` is wrong too.

**Comment by dannyyy** [comment]
@viceice The `endpoint` is according to the documentation on https://docs.renovatebot.com/nuget/. Also the `nuget` configuration property originates from that example. So I guess, the documentation is outdated?

The `packageRules` documentation on https://docs.renovatebot.com/configuration-options/#packagerules doesn't mention any word about which properties are allowed inside.

Is somewhere a schema available to get an idea of the current configuration possibilities? Most of the samples inside the documentation, issues and google-solutions are contradicting to each other. It would be awesome to evolve from a try & error  approach to a versioned schema approach.

**Comment by viceice** [comment]
@dannyyy  Do my suggestions work?

Yes, the docs seem to be outdated, we'll fix this soon. We have a json schema, you need a schema aware editor like vscode.

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  ...
}
```

@rarkins @HonkingGoose  We need to extend the packageRules and fix the nuget samples

**Comment by HonkingGoose** [comment]
I've made a new issue at #7959 to discuss the docs changes.

**Comment by dannyyy** [comment]
> @dannyyy Do my suggestions work?
> 
> Yes, the docs seem to be outdated, we'll fix this soon. We have a json schema, you need a schema aware editor like vscode.
> 
> ```json
> {
>   "$schema": "https://docs.renovatebot.com/renovate-schema.json",
>   ...
> }
> ```

In the mean time I had the chance to check your suggestions. Yes, all of them do. Thank you very much!

**Comment by rarkins** [comment]
Is there anything remaining to be done or can this issue be closed?

**Comment by viceice** [comment]
I think this should be moved to discussion's. I don't see any bug or feature here.
